keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/ResourceActuator.kt

```diff
@@ -304,31 +312,43 @@ class ResourceActuator(
       else -> SystemException(this)
     } as SpinnakerException
 
-  private suspend fun <T : Any> ResourceHandler<*, T>.resolve(resource: Resource<ResourceSpec>): Pair<T, T?> =
-    supervisorScope {
-      val desired = async {
-        try {
-          desired(resource)
-        } catch (e: ResourceCurrentlyUnresolvable) {
-          throw e
-        } catch (e: Throwable) {
-          throw CannotResolveDesiredState(resource.id, e)
+  private suspend fun <T : Any> ResourceHandler<*, T>.resolve(resource: Resource<ResourceSpec>): Pair<T, T?> {
+    val isKotlin = this.javaClass.isKotlinClass
+    return supervisorScope {
+      val desired = if (isKotlin) {
+        async {
+          try {
+            desired(resource)
+          } catch (e: ResourceCurrentlyUnresolvable) {
+            throw e
+          } catch (e: Throwable) {
+            throw CannotResolveDesiredState(resource.id, e)
+          }
         }
+      } else {
+        // for Java compatibility
+        desiredAsync(resource, asyncExecutor).asDeferred()
```

Comment:
This works as expected with respect to coroutine timeouts from outer scopes. The only downside I've experienced is that cancelling the `CompletableFuture` does *not* interrupt the underlying running thread(s), which is by design since the interface is entirely decoupled from how the future is processed (may even be synchronous). I think that's acceptable, considering we're "enforcing" the use of a managed thread pool and that we're planning to implement timeouts in the Java plugin as well. Let me know what you think.

Reply:
Good point on metrics. Not sure if there are any hooks available to instrument that (but now we have New Relic! 😉). BTW, if plugins are generally well-behaved, the cached thread pool should improve resource utilization, but they're not, we may end up with a lot of threads. We could switch that to a fixed thread pool instead. Thoughts?

Reply:
I'd leave it as-is for now, we can change the implementation to a fixed thread pool later if we run into performance issues.